src/matrix.rs

```diff
@@ -1,13 +1,43 @@
 use vector::{Vector2, Vector3, Vector4};
 use std::{mem, ptr};
 
+macro_rules! decl_matrix_type {
+    (
+        $(#[$attr:meta])*
+        $v:vis struct $name:ident {
+            $(
+                #[doc = $field_doc:expr]
+                $field_vis:vis $field:ident: $field_ty:ident
+            ),*
+        }
+    ) => {
+        $(#[$attr])*
+        $v struct $name<T> {
+            $(
+                #[doc = $field_doc]
+                $field_vis $field: $field_ty<T>
+            ),*
+        }
+    };
+}
+
 macro_rules! matrix {
-    ($name:ident : $vec:ident[ $($field:ident = $index:expr),* ] = ($inner:expr, $outer:expr)) => {
-        #[derive(Clone, Copy, Debug, Hash, PartialEq, PartialOrd, Eq, Ord)]
-        #[repr(C)]
-        #[allow(missing_docs)] //TODO: actually have docs
-        pub struct $name<T> {
-            $( pub $field : $vec<T>, )*
+    (
+        $( #[doc = $doc:expr] )*
+        $name:ident: $major_ty:ident $vec:ident[ $($field:ident = $index:expr),* ] = ($inner:expr, $outer:expr)
+    ) => {
+        decl_matrix_type! {
```

Comment:
is there a point for a separate macro here if it's not used elsewhere?

Reply:
So, documentation in macro types requires you to use the [`doc` attribute](https://doc.rust-lang.org/rustdoc/the-doc-attribute.html), which is quite unwieldy.

The attribute itself can only take either a string literal, or a `$expr` string as its single argument, which means that:

* you either have to construct the doc comment line by line using a sequence of `#[doc
 = "blah"] statements for interpolation, or:
* you have to do the interpolation using `concat!` and `stringify!`, and then pass that off to another macro ([example taken from here](https://stackoverflow.com/questions/43353077/how-to-embed-a-rust-macro-variable-into-documentation)), à la a c-style stringification macro.